pbt-gen/src/time.rs

```diff
@@ -0,0 +1,182 @@
+//! [proptest](https://github.com/AltSysrq/proptest) generators for time-like
+//! objects.
+
+use std::convert::TryInto;
+
+use chrono::{DateTime, NaiveDate, TimeZone, Timelike, Utc};
+use proptest::prelude::*;
+
+/// Any higher, and we're at seconds
+pub const MAX_NANO_SECS: u32 = 999_999_999u32;
+
+/// The most distant time in the past for which chrono produces correct
+/// times from [Utc.timestamp](chrono::Utc.timestamp).
+///
+/// See https://github.com/chronotope/chrono/issues/537.
+///
+/// ```
+/// use pbt_gen;
+///
+/// assert_eq!(pbt_gen::time::min_time().to_string(), "1653-02-10 06:13:21 UTC".to_string());
+/// ```
+pub fn min_time() -> DateTime<Utc> {
+    Utc.timestamp(-9999999999, 0)
+}
+
+/// The most distant time in the future for which chrono produces correct
+/// times from [Utc.timestamp](chrono::Utc.timestamp).
+///
+/// See https://github.com/chronotope/chrono/issues/537.
+///
+/// ```
+/// use pbt_gen;
+///
+/// assert_eq!(pbt_gen::time::max_time().to_string(), "5138-11-16 09:46:39 UTC".to_string());
+/// ```
+pub fn max_time() -> DateTime<Utc> {
+    Utc.timestamp(99999999999, 0)
+}
+
+fn num_days_in_month(year: i32, month: u32) -> u32 {
+    // Using chrono, we get the duration beteween this month and the next,
+    // then count the number of days in that duration.  See
+    // https://stackoverflow.com/a/58188385/1187277
+    let given_month = NaiveDate::from_ymd(year, month, 1);
+    let next_month = NaiveDate::from_ymd(
+        if month == 12 { year + 1 } else { year },
+        if month == 12 { 1 } else { month + 1 },
+        1,
+    );
+    next_month
+        .signed_duration_since(given_month)
+        .num_days()
+        .try_into()
+        .unwrap()
+}
+
+prop_compose! {
+    /// An abitrary [chrono::DateTime] that is between the given `min`
+    /// and `max`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use chrono::{TimeZone, Utc};
+    /// use pbt_gen;
+    /// use proptest::prelude::*;
+    ///
+    /// proptest!{
+    ///     fn rosa_luxemburg_and_octavia_butler_were_not_alive_at_the_same_time(
+    ///        time_in_luxemburgs_lifespan in pbt_gen::time::arb_datetime_in_range(
+    ///          Utc.ymd(1871, 3, 5).and_hms(0,0,0), // DOB
+    ///          Utc.ymd(1919, 1, 15).and_hms(0,0,0), // DOD
+    ///        ),
+    ///        time_in_butlers_lifespan in pbt_gen::time::arb_datetime_in_range(
+    ///          Utc.ymd(1947, 6, 22).and_hms(0,0,0), // DOB
+    ///          Utc.ymd(2006, 2, 24).and_hms(0,0,0), // DOD
+    ///        ),
+    ///     ) {
+    ///       prop_assert!(time_in_butlers_lifespan == time_in_butlers_lifespan)
```

Comment:
isn't this expected to be not equal?
```suggestion
    ///       prop_assert!(time_in_butlers_lifespan != time_in_butlers_lifespan)
```

Reply:
Indeed! I messed this up. I also put the same lifespan twice, I now see. Thanks for catching my mistake!

Reply:
Ouch, I didn't see that 😄